tests/auto/tst_persistenttabmodel/tst_persistenttabmodel.cpp

```diff
@@ -0,0 +1,424 @@
+        "Item {\n" \
+        "   width: 100; height: 100\n" \
+        "   property alias tabModel: model\n" \
+        "   PersistentTabModel { id: model }\n" \
+        "}\n";
+
+class tst_persistenttabmodel : public QObject
+{
+    Q_OBJECT
+
+private slots:
+    void initTestCase();
+    void cleanupTestCase();
+    void init();
+    void cleanup();
+
+    void addTab();
+    void remove();
+    void removeTabById_data();
+    void removeTabById();
+    void clear();
+    void activateTab_by_url_data();
+    void activateTab_by_url();
+    void activateTabById_data();
+    void activateTabById();
+    void closeActiveTab();
+    void updateUrl_data();
+    void updateUrl();
+    void onTitleChanged();
+
+private:
+    void addThreeTabs();
+
+    PersistentTabModel* tabModel;
+    QQuickView mView;
+    QPointer<QQmlComponent> mComponent;
+    QObject* mRootObject;
+    QString mDbFile;
+};
+
+void tst_persistenttabmodel::initTestCase()
+{
+    qGuiApp->setAttribute(Qt::AA_Use96Dpi, true);
+    qmlRegisterUncreatableType<DeclarativeTabModel>("Sailfish.Browser", 1, 0, "TabModel",
+                                                    "TabModel is abstract!");
+    qmlRegisterType<PersistentTabModel>("Sailfish.Browser", 1, 0, "PersistentTabModel");
+    mComponent = new QQmlComponent(mView.engine());
+    mComponent->setData(QML_SNIPPET, QUrl());
+    mDbFile = QString("%1/%2")
+            .arg(QStandardPaths::writableLocation(QStandardPaths::DataLocation))
+            .arg(QLatin1String(DB_NAME));
+}
+
+void tst_persistenttabmodel::cleanupTestCase()
+{
+    delete mComponent;
+}
+
+void tst_persistenttabmodel::init()
+{
+    mRootObject = mComponent->create(mView.engine()->rootContext());
+    QVariant var = mRootObject->property("tabModel");
+    tabModel = qobject_cast<PersistentTabModel*>(qvariant_cast<QObject*>(var));
+    QVERIFY(tabModel);
+
+    if (!tabModel->loaded()) {
+        QSignalSpy loadedSpy(tabModel, SIGNAL(loadedChanged()));
+        // Tabs must be loaded with in 500ms
+        QVERIFY(loadedSpy.wait());
+        QCOMPARE(loadedSpy.count(), 1);
+    }
+}
+
+void tst_persistenttabmodel::cleanup()
+{
+    delete mRootObject;
+    delete DBManager::instance();
+    QFile dbFile(mDbFile);
+    QVERIFY(dbFile.remove());
+}
+
+void tst_persistenttabmodel::addTab()
+{
+    QSignalSpy countChangeSpy(tabModel, SIGNAL(countChanged()));
+    QSignalSpy tabAddedSpy(tabModel, SIGNAL(tabAdded(int)));
+    QSignalSpy nextTabIdChangedSpy(tabModel, SIGNAL(nextTabIdChanged()));
+    QSignalSpy dataChangedSpy(tabModel, SIGNAL(dataChanged(QModelIndex, QModelIndex, QVector<int>)));
+    QSignalSpy activeTabIndexChangedSpy(tabModel, SIGNAL(activeTabIndexChanged()));
+    QSignalSpy activeTabChangedSpy(tabModel, SIGNAL(activeTabChanged(int,bool)));
+
+    QList<QString> urls, titles;
+    urls << "http://example.com" << "file:///opt/tests/testpage.html" << "https://example.com";
+    titles << "Test title1" << "Test title2" << "Test title3";
```

Comment:
My point was that this could be changed so...

```
    QTest::addColumn<int>("insertToIndex");
    QTest::addColumn<QString>("newUrl");
    QTest::addColumn<QString>("newTitle");
    QTest::addColumn<QList<TestTab>("expectedTabs");
    // Each tab data and expected count would be covered by the above. Also model order could be checked.
```
TestTab could be moved from the testobject.h to own header and cpp and changed to be a QObject if needed.

In addition, it could be extended to include also tabId.